### PR TITLE
[WSJ] Extract videos embedded in articles, too

### PR DESCRIPTION
**Pull request by johnhawkinson**
- [x] Improvement

### CODE REVIEW

File: youtube_dl/extractor/wsj.py [hunk 1]
```diff
@@ -87,3 +89,24 @@ def _real_extract(self, url):
...
+            'uploader_id': 'ralcaraz',
+            'title': 'Bao Bao the Panda Leaves for China',
+        }
+    }]
+
+    def _real_extract(self, url):
+        article_id = self._match_id(url)
+        webpage = self._download_webpage(url, article_id)
+        video_id = self._html_search_regex(r'data-src=["\']([A-Z0-9\-]+)',
+                                           webpage, 'video id')
```
Comment (dstftw) [inline youtube_dl/extractor/wsj.py]: `_search_regex` is enough here.

File: youtube_dl/extractor/wsj.py [hunk 2]
```diff
@@ -87,3 +89,24 @@ def _real_extract(self, url):
...
+        return self._extract_video(video_id)
```
Comment (dstftw) [inline youtube_dl/extractor/wsj.py]: Instead if this extraction should be delegated to `WSJIE` via `url_result`.

File: youtube_dl/extractor/wsj.py [hunk 3]
```diff
@@ -17,6 +17,7 @@ class WSJIE(InfoExtractor):
         )
         (?P<id>[a-zA-Z0-9-]+)'''
     IE_DESC = 'Wall Street Journal'
+    TEMPLATE_URL = 'https://wsj.com/video/./%s'
```
Comment (dstftw) [inline youtube_dl/extractor/wsj.py]: What's the point extracting it here? Also use http://video-api.wsj.com/api-video/player/iframe.html?guid=... instead or add a `wsj:` shortcut.

File: youtube_dl/extractor/wsj.py [hunk 4]
```diff
@@ -87,3 +88,24 @@ def _real_extract(self, url):
             'title': title,
             'categories': info.get('keywords'),
         }
+
+
+class WSJArticleIE(WSJIE):
```
Comment (dstftw) [inline youtube_dl/extractor/wsj.py]: There is no point inheriting from `WSJIE`.

File: youtube_dl/extractor/wsj.py [hunk 5]
```diff
@@ -87,3 +88,24 @@ def _real_extract(self, url):
...
+            'title': 'Bao Bao the Panda Leaves for China',
+        }
+    }]
+
+    def _real_extract(self, url):
+        article_id = self._match_id(url)
+        webpage = self._download_webpage(url, article_id)
+        video_id = self._search_regex(r'data-src=["\']([A-Z0-9\-]+)',
+                                      webpage, 'video id')
+        return self.url_result(self.TEMPLATE_URL % video_id, WSJIE.ie_key())
```
Comment (dstftw) [inline youtube_dl/extractor/wsj.py]: Also pass `video_id` since it's known beforehand.

**Comment by johnhawkinson** [comment]
Sorry, I was doing some research for this comment when you typed yours, so this is kind of duplicative.

> What's the point extracting it here?

Sorry, I don't understand the question. But maybe it's irrelevent with a `wsj:` shortcut.

---

> Instead of this extraction should be delegated to WSJIE via url_result.

Err...I'm definitely not sure that I did this right.
It kind of feels like introducing a `wsj:` pseudoURL  is overkill here. But maybe I should have done that instead? **Clearly.**

I added a template but unfortunately:

```
    _VALID_URL = r'''(?x)https?://
        (?:
            video-api\.wsj\.com/api-video/player/iframe\.html\?guid=|
            (?:www\.)?wsj\.com/video/[^/]+/
        )
        (?P<id>[a-zA-Z0-9-]+)'''
    TEMPLATE_URL = 'https://wsj.com/video/./%s'
```

It really doesn't seem like it's appropriate to put anything at all in first field, so I went with `/./`; and constructing the iframe URL sees kind of wrong, too.

Incidently, I also wasn't really sure how to name this template function, there seems to be a lot of variety:

```
pb3:extractor jhawk$ gsed -n -e 's/^\s*//' -e '/TEMPLATE[A-Z_]* *=/s/=.*//p' *py| sort | uniq -c | sort -nr
  14 BRIGHTCOVE_URL_TEMPLATE 
  12 _URL_TEMPLATE 
   6 _API_URL_TEMPLATE 
   5 _TEMPLATE_URL 
   5 _PAGE_TEMPLATE 
   2 _VIDEO_PAGE_TEMPLATE 
   2 _VIDEO_ID_TEMPLATE 
   2 _VALID_URL_TEMPLATE 
   2 _MOBILE_TEMPLATE 
   1 _VIDEO_INFO_TEMPLATE 
   1 _URL_VIDEO_TEMPLATE 
   1 _THUMBNAIL_URL_TEMPLATE 
   1 _SUB_LANG_TEMPLATE 
   1 _SONG_URL_TEMPLATE 
   1 _SONG_FILE_URL_TEMPLATE 
   1 _SERVICE_PROVIDER_TEMPLATE 
   1 _SECURITY_URL_TEMPLATE 
   1 _PLAYER_URL_TEMPLATE 
   1 _PAGE_DATA_TEMPLATE 
   1 _MOVIE_TEMPLATE 
   1 _EPISODE_URL_TEMPLATE 
   1 _AUTHORIZATION_URL_TEMPLATE 
   1 _API_QUERY_TEMPLATE 
   1 _API_BASE_TEMPLATE 
   1 _AGECHECK_TEMPLATE 
   1 TP_RELEASE_URL_TEMPLATE 
   1 TEMPLATE_URL 
   1 SOAP_TEMPLATE 
   1 PROGRESSIVE_URL_TEMPLATE 
   1 PG_URL_TEMPLATE 
   1 META_DATA_URL_TEMPLATE 
   1 ADAPTIVE_URL_TEMPLATE 
pb3:extractor jhawk$ 
```

I guess the one takeaway is I should use a leading underscore. I really don't understand why youtube-dl uses them so much, they seem to add a lot of confusion and (not here) make docstrings not work.

If only there were a "A model, well-commented extractor" to look at :) 
#11968

**Pull request closed without merging**